File: diffbotapi.py
import requests
import json
import datetime
from secret import DIFFBOT_API_TOKEN


def query_dql(subject, token=DIFFBOT_API_TOKEN):
    '''
        Parameters: Subject of query, optional token
        Return:
        API call for articles on subject
    '''
    today = datetime.date.today()

    # Generate dates from the start of the week to today
    week = [today - datetime.timedelta(days=i) for i in range(7)]

    querystring = {
        'token': token,
        'query': f'type:Article date<"{week[0]}" date>"{week[-1]}" title:"{subject}" language:"en" sortBy:date',
        'format': "json", 
        'size': 2       
    }
    response = requests.get('https://kg.diffbot.com/kg/v3/dql', params=querystring)
    data = response.json()
    articles = {}
    for i in range(len(data.get('data'))):
        article = data.get('data')[i].get('entity')
        articles[article.get('pageUrl')] = {'title' : article.get('title'), 'text' : article.get('text'), 'classification' : article.get('categories')}
    return articles
    
'''
def fetch_article(url, token=DIFFBOT_API_TOKEN):
    """
        Parameters: url of article, optional token
        Return:
        the url of the article
        title of the article
        categories of which the article falls for
        text of article
    """
    encoded_url = requests.utils.quote(url)

    request_url = f"https://api.diffbot.com/v3/article?url={encoded_url}&token={token}"

    headers = {"accept": "application/json"}

    response = requests.get(request_url, headers=headers)
 
    json_content = response.json()

    title = json_content['objects'][0]['title']

    if 'categories' not in json_content['objects'][0]:
        catergory = json_content['objects'][0]['categories']

        classification = [name for name in catergory]
    else:
        tags = json_content['objects'][0]['tags']
        tags = [(tag['count'], tag['name']) for tag in tags]
        # find top 3 tags
        classification = sorted(tags, key=lambda x: x[0], reverse=True)[:3]



    text = json_content['objects'][0]['text']
    return url, title, classification, text
'''
response = query_dql('bitcoin')
print(response)

File: my-project/src/backend/diffbotapi.js
// src/backend/diffbotapi.js
import axios from "axios";
import { summarizeText } from "./cohereaiapi.js";
import dotenv from "dotenv";
dotenv.config({ path: "./src/keys.env" });
const apiKey = process.env.DIFFBOTAPI_TOKEN;

async function queryDQL(subject, token = apiKey) {
  const today = new Date();
  const TodayDate = new Date(
    today.getTime() - today.getTimezoneOffset() * 60000,
  )
    .toISOString()
    .split("T")[0];

  const querystring = {
    token: token,
    query: `type:Article date:"${TodayDate}" title:"${subject}" language:"en" sortBy:date`,
    format: "json",
    size: 4, // Retrieve 5 articles
  };

  try {
    const response = await axios.get("https://kg.diffbot.com/kg/v3/dql", {
      params: querystring,
    });
    const data = response.data;
    const articles = [];
    if (data && data.data) {
      for (const articleEntity of data.data) {
        const article = articleEntity.entity;
        if (article && article.text) {
          // Ensure the text is treated as a string and summarize it
          const text = String(article.text); // Converts non-string input to string
          const summary = await summarizeText(text);
          articles.push({
            url: article.pageUrl,
            title: article.title,
            summary: summary,
          });
        }
      }
    }
    return articles;
  } catch (error) {
    console.error("Error querying DQL:", error);
    return [];
  }
}

export { queryDQL };
